articles/151_roslaunch_xml.md

```diff
@@ -0,0 +1,744 @@
+      <xs:annotation>
+        <xs:documentation xml:lang="en">
+          Name of the node executable.
+        </xs:documentation>
+      </xs:annotation>
+    </xs:attribute>
+    <xs:attribute name="name" type="xs:string" use="optional">
+      <xs:annotation>
+        <xs:documentation xml:lang="en">
+          A name for the launched ROS node.
+        </xs:documentation>
+      </xs:annotation> 
+    </xs:attribute>
+      <xs:attribute name="args" type="xs:string" use="optional">
+        <xs:annotation>
+          <xs:documentation xml:lang="en">
+            Additional 'command-line' arguments for the ROS node.
+          </xs:documentation>
+        </xs:annotation>
+      </xs:attribute>
+      <xs:attribute name="ns" type="xs:string" use="optional">
+        <xs:annotation>
+          <xs:documentation xml:lang="en">
+            A ROS namespace to scope the launched ROS node.
+          </xs:documentation>
+        </xs:annotation>
+      </xs:attribute>
+      <xs:attribute name="launch-prefix" type="xs:string" use="optional">
+        <xs:annotation>
+          <xs:documentation xml:lang="en">
+            A prefix for the ROS node command-line used for launch.
+          </xs:documentation>
+        </xs:annotation>
+      </xs:attribute>
+      <xs:attribute name="output" use="optional">
+        <xs:annotation>
+          <xs:documentation xml:lang="en">
+            Output type for the launched ROS node.
+          </xs:documentation>
+        </xs:annotation>
+        <xs:simpleType>
+          <xs:restriction base="xs:string">
+            <xs:enumeration value="log">
+              <xs:annotation>
+                <xs:documentation xml:lang="en">
+                  ROS node output goes to a log file.
+                </xs:documentation>
+              </xs:annotation>
+            </xs:enumeration>
+            <xs:enumeration value="screen">
+              <xs:annotation>
+                <xs:documentation xml:lang="en">
+                  ROS node output goes to stdout.
+                </xs:documentation>
+              </xs:annotation>
+            </xs:enumeration>
+          </xs:restriction>
+        </xs:simpleType>
+      </xs:attribute>
+      <xs:attribute name="if" type="xs:string" use="optional">
+        <xs:annotation>
+          <xs:documentation xml:lang="en">
+            A predicate to condition ROS node launch i.e. only do
+            so if the predicate evaluates to a truthy value.
+          </xs:documentation>
+        </xs:annotation>
+      </xs:attribute>
+      <xs:attribute name="unless" type="xs:string" use="optional">
+        <xs:annotation>
+          <xs:documentation xml:lang="en">
+            A predicate to condition ROS node launch i.e. do so unless
+            the predicate evaluates to a truthy value.
+          </xs:documentation>
+        </xs:annotation>
+      </xs:attribute>
+  </xs:element>
+</xs:schema> 
+```
```

Comment:
I fully understand your concerns, and I expect this to change over time as well. But being the up-to-date description of the format is IMHO the sole purpose of this document in the long term.

As we add markup for more actions, substitutions, etc., I think we have to consider the possibility that not all ROS 2 launch implementations will support all of them (no interest, no resources, too difficult to implement, etc.) Also, if we enable users to develop and integrate their own, I think we have to define what's standard and what's not. 

In that context, this becomes an statement: for a ROS 2 launch implementation to support XML launch v0.1.0 it has to implement enough functionality for a description compliant with v0.1.0 to work. Anything outside of it is an extension.

Reply:
I'm in favor of having and maintaining a regersious XML schema definition that enforces the launch specification. Where such a `.xst` document should reside is perhaps the notable question. Having the `.xst` schema file in close proximity/revision control with design description could help ensure they stay in sync, but how does ROS1 related REPs host/track the schema definitions for ros1 launch.xml and package.xml files?

Reply:
The package manifest formats do come with a `.xsd` schema file which is referenced in the REP (e.g. [REP 149](http://www.ros.org/reps/rep-0149.html#schema)), stored in the `rep` repo, and hosted on http://download.ros.org/schema/ to be referenced by `package.xml` files.

I don't think an official schema for ROS 1 launch file exists at the moment.

Reply:
Perhaps having the schema definition in it own `.xsd` (sorry, I meant this extension originally) would be better than inlining it in this markdown document, so commit history for the Schema Definition itself would be a bit cleaner.

Reply:
Alright, I moved the schema to its own `.xsd` file and I managed to get Jekyll to include it verbatim in the format specification document. 

@wjwwood @dirk-thomas what do you think of this addition to the site layout I'm making here?

Reply:
I think this will help avoid needing to keep multiple files in sync, so I'm a fan of it. I don't know how we keep the copy on downloads.ros.org in sync with the one in the rep repostiory, but we'll have to do something similar here, unless we host directly from design.ros2.org, which isn't a great idea I think.

Reply:
The `xsd` file itself should be versioned. To the URL used by launch files to validate is "fixed" and the `xsd` file referenced by it does never change. Therefore it is a one time upload to downloads.tos.org - no need to keep anything in sync since the `xsd` file isn't supposed to change. Only a newer version should be uploaded with a new name / URL.